cmake/match.py

```diff
@@ -5,39 +5,137 @@
+#
 
 import sys
 import re
+from enum import Enum
 
-if len(sys.argv) != 3:
-    print("Usage: python match.py <input_file> <match_file>")
-    sys.exit(1)
 
-input_file = sys.argv[1]
-match_file = sys.argv[2]
+## @brief print the whole content of input and match files
+def print_content(input_lines, match_lines, ignored_lines):
+    print("--- Input Lines " + "-" * 64)
+    print("".join(input_lines).strip())
+    print("--- Match Lines " + "-" * 64)
+    print("".join(match_lines).strip())
+    print("--- Ignored Lines " + "-" * 62)
+    print("".join(ignored_lines).strip())
+    print("-" * 80)
 
-with open(input_file, 'r') as input, open(match_file, 'r') as match:
-    input_lines = input.readlines()
-    match_lines = match.readlines()
 
-if len(match_lines) != len(input_lines):
-    sys.exit(f"Line count doesn't match (is: {len(input_lines)}, expected: {len(match_lines)})")
+def print_incorrect_match(match_line, present, expected):
+    print("Line " + str(match_line) + " does not match")
+    print("is:       " + present)
+    print("expected: " + expected)
+
+
+## @brief pattern matching script status values
+class Status(Enum):
+    INPUT_END = 1
+    MATCH_END = 2
+    INPUT_AND_MATCH_END = 3
+    PROCESSING = 4
+
+
+## @brief check matching script status
+def check_status(input_lines, match_lines):
+    if not input_lines and not match_lines:
+        return Status.INPUT_AND_MATCH_END
+    elif not input_lines:
+        return Status.INPUT_END
+    elif not match_lines:
+        return Status.MATCH_END
+    return Status.PROCESSING
 
-for i, match_line in enumerate(match_lines):
-    # split into parts at {{ }}
-    match_parts = re.split(r'\{{(.*?)\}}', match_line.strip())
-    pattern = ""
-    for j, part in enumerate(match_parts):
-        if j % 2 == 0:
-            pattern += re.escape(part)
-        else:
-            pattern += part
 
-    input_line = input_lines[i].strip()
-    if not re.fullmatch(pattern, input_line):
-        print(f"Line {i+1} does not match".format(i+1))
-        print("is:       " + input_line)
-        print("expected: " + match_line.strip())
+## @brief pattern matching tags.
+## Tags are expected to be at the start of the line.
+class Tag(Enum):
+    OPT = "{{OPT}}"         # makes the line optional
+    IGNORE = "{{IGNORE}}"   # ignores all input until next match or end of input file
+
+
+## @brief main function for the match file processing script
+def main():
+    if len(sys.argv) != 3:
+        print("Usage: python match.py <input_file> <match_file>")
         sys.exit(1)
+
+    input_file = sys.argv[1]
+    match_file = sys.argv[2]
+
+    with open(input_file, 'r') as input, open(match_file, 'r') as match:
+        input_lines = input.readlines()
+        match_lines = match.readlines()
+
+    ignored_lines = []
+
+    input_idx = 0
+    match_idx = 0
+    tags_in_effect = []
+    status = Status.PROCESSING
+    while status == Status.PROCESSING:
```

Comment:
this doesn't seem necessary, this condition will never be false as far as I can tell. You can just do a while True: loop and check status at the end (and exit appropriately).

Reply:
I've moved the status variable inside the loop and changed the loop condition.
If I were to move the status check at the end, then I would need to check if either input or match file is empty to avoid index error. Logger has such test where both input and match files are empty.

Reply:
Also, the match case statement (switch equivalent) is not available in our current supported python version.